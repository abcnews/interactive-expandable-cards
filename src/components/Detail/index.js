const classNames = require('classnames');
const { h, Component } = require('preact');
const styles = require('./styles.scss');

const TABBABLE_SELECTOR = '[href], button, input:not([type="hidden"]), select, textarea, [tabindex]';

class Detail extends Component {
  constructor(props) {
    super(props);

    this.toggleTabbable = this.toggleTabbable.bind(this);
    this.getContentRef = this.getContentRef.bind(this);
  }

  getContentRef(el) {
    this.content = el;
  }

  animateHeightChange() {
    this.base.style.height = `${this.content.clientHeight}px`;
    setTimeout(() => {
      this.base.style.height = this.props.open ? 'auto' : '0';
    }, this.props.open ? 250 : 0);
  }

  toggleTabbable() {
    this.tabbable.forEach(x => {
      if (this.props.open) {
        if (x.initial === null) {
          x.el.removeAttribute('tabindex');
        } else {
          x.el.setAttribute('tabindex', x.initial);
        }
      } else {
        x.el.setAttribute('tabindex', -1);
      }
    });
  }

  componentDidMount() {
    this.props.nodes.forEach(node => {
      this.content.appendChild(node);
    });

    this.tabbable = [...this.content.querySelectorAll(TABBABLE_SELECTOR)].map(el => ({
      el,
      initial: el.getAttribute('tabindex')
    }));

    this.animateHeightChange();
    this.toggleTabbable();
  }

  componentDidUpdate() {
    this.animateHeightChange();
    this.toggleTabbable();
  }

  render() {
    return (
      <div className={classNames(styles.root, { [styles.open]: this.props.open })} data-component="Detail">
        <div ref={this.getContentRef} className={`${styles.content} u-richtext`} />
      </div>
    );
  }
}

module.exports = Detail;
