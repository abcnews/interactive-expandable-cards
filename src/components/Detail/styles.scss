.root {
  overflow: hidden;
  margin: 0 0.625rem;
  font-family: ABCSans, Helvetica, Arial, sans-serif;

  @media (min-width: 40rem) {
    opacity: 0;
    transform: translate(0, -1.5rem);
    position: relative;
    top: -3.5rem;
    margin-left: 0;
    margin-right: 0;
    margin-bottom: -1.5rem;
    transition: opacity 0.25s cubic-bezier(0.645, 0.045, 0.355, 1), transform 0.25s cubic-bezier(0.645, 0.045, 0.355, 1),
      height 0.25s cubic-bezier(0.645, 0.045, 0.355, 1);

    > div {
      margin-left: 0;
      margin-right: 0;
      padding-bottom: 2.5rem;
    }
  }

  @media (prefers-reduced-motion) {
    transition: none;
  }
}

.open {
  @media (min-width: 40rem) {
    opacity: 1;
    transform: none;
  }
}

.content {
  padding: 1rem 0 1.25rem;

  @media (min-width: 40rem) {
    padding: 5rem 2.5rem 4.5rem;
  }
  & > :first-child {
    margin-top: 0 !important;
  }

  & > :last-child {
    margin-bottom: 0 !important;
  }
}

.image {
  display: block;
  width: 100%;
  height: auto;
}
