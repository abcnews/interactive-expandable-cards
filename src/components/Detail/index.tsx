import classNames from 'classnames';
import { h } from 'preact';
import { useRef, useState, useEffect, useLayoutEffect } from 'preact/hooks';
import styles from './styles.scss';

type DetailProps = {
  open: boolean;
  nodes: HTMLElement[];
};

const TABBABLE_SELECTOR = '[href], button, input:not([type="hidden"]), select, textarea, [tabindex]';

export const Detail = ({ open, nodes }: DetailProps) => {
  const contentRef = useRef<HTMLDivElement>();
  const baseRef = useRef<HTMLDivElement>();
  const [tabbable, setTabbable] = useState<{ el: Element; initial: string | null }[]>([]);

  const animateHeightChange = () => {
    if (baseRef.current instanceof HTMLElement) {
      const el = baseRef.current;
      el.style.height = `${contentRef.current.clientHeight}px`;
      setTimeout(
        () => {
          el.style.height = open ? 'auto' : '0';
        },
        open ? 250 : 0
      );
    }
  };

  tabbable.forEach(x => {
    if (open) {
      if (x.initial === null) {
        x.el.removeAttribute('tabindex');
      } else {
        x.el.setAttribute('tabindex', x.initial);
      }
    } else {
      x.el.setAttribute('tabindex', '-1');
    }
  });

  useEffect(() => {
    nodes.forEach(node => {
      contentRef.current.appendChild(node);
    });

    setTabbable(
      Array.from(contentRef.current.querySelectorAll(TABBABLE_SELECTOR)).map(el => ({
        el,
        initial: el.getAttribute('tabindex')
      }))
    );

    return () => {
      while (contentRef.current.firstChild) {
        contentRef.current.firstChild.remove();
      }
      setTabbable([]);
    };
  }, [nodes]);

  useLayoutEffect(() => {
    animateHeightChange();
  }, [open]);

  return (
    <div ref={baseRef} className={classNames(styles.root, { [styles.open]: open })} data-component="Detail">
      <div ref={contentRef} className={`${styles.content} u-richtext`} />
    </div>
  );
};
