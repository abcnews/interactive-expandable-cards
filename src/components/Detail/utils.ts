import type { ExpandableCardsImage } from '../ExpandableCards';
import styles from './styles.scss';

export const createImage = (image: ExpandableCardsImage): HTMLImageElement => {
  const el = document.createElement('img');

  el.className = styles.image;
  el.alt = image.alt;
  el.srcset = image.renditions.map(i => `${i.url} ${i.width}w`).join(', ');

  return el;
};
