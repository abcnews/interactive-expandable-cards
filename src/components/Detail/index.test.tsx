import { h } from 'preact';
import render from 'preact-render-to-string';

import Detail from '.';

describe('Detail', () => {
  test('It renders', () => {
    const actual = render(<Detail open={false} nodes={[document.createElement('p')]} />);
    expect(actual).toMatchSnapshot();
  });
});
