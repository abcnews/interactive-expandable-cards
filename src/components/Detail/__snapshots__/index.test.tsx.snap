// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail It renders 1`] = `"<div class data-component=\\"Detail\\"><div class=\\"undefined u-richtext\\"></div></div>"`;
